prometheus_flask_exporter/__init__.py

```diff
@@ -662,6 +665,10 @@ def _track(self, metric_type, metric_call, metric_kwargs, name, description, lab
             **metric_kwargs
         )
 
+        # When all labels are already known at this point, the metric can get an initial value.
```

Comment:
OK, this surfaced a bunch of issues, so perhaps it can't be default on?
https://github.com/rycus86/prometheus_flask_exporter/actions/runs/3410937217/jobs/5688051579

We should also add some tests when it is on though, to check the behaviour there.
And if this becomes a flag, can I kindly ask you to also include a bit on it in the README?
Thank you!

Reply:
Looking at the tests again, I think it has an issue when you call into `parent_metric.labels` function with no labels present (e.g. the metric only has a name and description).

Reply:
Oops! Forgot to run the tests 😅

Ok I modified the code to check for if there are labels present and now the tests succeed.

I also added a few assertions to the tests that check if metric values are 0 at the start.

I understand that adding a flag is a bit safer because it basically skips the new code.
But it also complicates the configuration (and documentation).

So the question is:
Are the test now enough to show that the proposed code changes don't (accidentally) change other things?
Or is it still better to use a feature flag? And if so do you have an idea how to properly modify the tests?

Reply:
> Are the test now enough to show that the proposed code changes don't (accidentally) change other things?

It looks promising, I'm just unsure what is the implication of the initial values on the Prometheus end, for example a newly started instance would define a gauge for idk something like current temperature, and populate the value as soon as possible, but perhaps not immediately, then the new behaviour would report zero on it initially, while the old/current behaviour would just omit a metric. If there's no meaningful difference on Prometheus between these two scraped values, then perhaps we don't need the flag, I'd just feel more comfortable if people would have a way to either opt-out or opt-in to this new behaviour.

> And if so do you have an idea how to properly modify the tests?

Probably a test showing the default behaviour and the non-default behaviour with this is enough, the rest of the tests show that nothing is seriously wrong.
With tests, it would also be good to add a new test case that defines a metric with static labels, and shows that it gets the initial default value, and another metric with callable label values that does not get the default.

Reply:
> I'm just unsure what is the implication of the initial values on the Prometheus end, for example a newly started instance would define a gauge for idk something like current temperature, and populate the value as soon as possible, but perhaps not immediately, then the new behaviour would report zero on it initially, while the old/current behaviour would just omit a metric. If there's no meaningful difference on Prometheus between these two scraped values, then perhaps we don't need the flag, 

hmm, that's a good point.
For gauges it is less clear in some situatiuons what you want as an initial value.. or if you want an initial value at all.

I'm not sure if that is a problem for this library however, since the gauges in this library track the number of invocations in progress, which will be 0 at the start, just like the initial value.

>I'd just feel more comfortable if people would have a way to either opt-out or opt-in to this new behaviour.

I will add some documentation about when metrics will get an initial value, add an opt-out option, and write some tests that check that there is no initial value when using that option.